stdlib/dynarray.ml

```diff
@@ -102,17 +53,246 @@ and 'a slot =
+        | Some v -> ...]}
+      can instead be written
+      {[if Dummy.is_dummy x
+        then ...
+         else let v = Dummy.unsafe_get x in ...]}
+  *)
+
+  (** {4 Arrays of values or dummies} *)
+  module Array : sig
+    val make :
+      int -> 'a -> dummy:'stamp dummy ->
+      ('a, 'stamp) with_dummy array
+
+    val init :
+      int -> (int -> 'a) -> dummy:'stamp dummy ->
+      ('a, 'stamp) with_dummy array
+
+    val copy : 'a array -> dummy:'stamp dummy -> ('a, 'stamp) with_dummy array
+
+    val unsafe_nocopy :
+      'a array -> dummy:'stamp dummy ->
+      ('a, 'stamp) with_dummy array
+    (** [unsafe_nocopy] assumes that the input array was created
+        locally and will not be used anymore (in the spirit of
+        [Bytes.unsafe_to_string]), and avoids a copy of the input
+        array when possible. *)
+
+    val blit_array :
+      'a array -> int ->
+      ('a, 'stamp) with_dummy array -> int ->
+      len:int ->
+      unit
+    val prefix :
+      ('a, 'stamp) with_dummy array ->
+      int ->
+      ('a, 'stamp) with_dummy array
+    val extend :
+      ('a, 'stamp) with_dummy array ->
+      length:int ->
+      dummy:'stamp dummy ->
+      new_capacity:int ->
+      ('a, 'stamp) with_dummy array
+  end
+end = struct
+  (* We want to use a cyclic value so that No_sharing marshalling
+     fails loudly, but we want also comparison of dynarrays to work
+     as expected, and not loop forever.
+
+     Our approach is to use an object value that contains a cycle.
+     Objects are compared by their unique id, so comparison is not
+     structural and will not loop on the cycle, but marshalled
+     by content, so marshalling without sharing will fail on the cycle.
+
+     (It is a bit tricky to build an object that does not contain
+     functional values where marshalling fails, see [fresh ()] below
+     for how we do it.) *)
+  type 'stamp dummy = < >
+  type fresh_dummy = Fresh : 'stamp dummy -> fresh_dummy
+
+  let fresh () =
+    (* dummies and marshalling: we intentionally
+       use a cyclic value here. *)
+    let r = ref None in
+    let dummy = object
+      val x = r
+    end in
+    r := Some dummy;
+    Fresh dummy
+
+  type ('a, 'stamp) with_dummy = 'a
+
+  let of_val v = v
+
+  let of_dummy (type a stamp) (dummy : stamp dummy) =
+    (Obj.magic dummy : (a, stamp) with_dummy)
+
+  let is_dummy v dummy =
+    v == of_dummy dummy
+
+  let unsafe_get v =
+    v
+
+  module Array = struct
+    let make n x ~dummy =
+      if Obj.(tag (repr x) <> double_tag) then
+        Array.make n (of_val x)
+      else begin
+        let arr = Array.make n (of_dummy dummy) in
+        Array.fill arr 0 n (of_val x);
+        arr
+      end
+
+    let copy a ~dummy =
+      if Obj.(tag (repr a) <> double_array_tag) then
+        Array.copy a
+      else begin
+        let n = Array.length a in
+        let arr = Array.make n (of_dummy dummy) in
+        Array.blit a 0 arr 0 n;
+        arr
+      end
```

Comment:
Ah, so it looks like you found a bug due to incorrect assumptions about `Array.blit`.

I'm not sure that the explanation of the bug is correct: the implementation of `blit` in [runtime/array.c:caml_array_blit](https://github.com/ocaml/ocaml/blob/d0ac496c2dc3e174f2bfff8ad78da1f398c24921/runtime/array.c#L382) does not make the assumption that you describe -- or at least I am not seeing it in the code. In the code I see a check on whether the destination array is a float array, but no check on the source array. I think that was is going on is that the non-flat-float-array implementation is incorrect if the source array is a flat float array.

In any case, your conclusion is correct that I should not use `Array.blit` here, I will update the code. Thanks!

Reply:
>  I think that was is going on is that the non-flat-float-array implementation is incorrect if the source array is a flat float array.

Ah yes, you’re right. I would complete: if the source array is a flat float array and the destination array is not, which of course is not supposed to happen.

Reply:
I pushed a new commit that should fix this issue by avoiding all uses of `Array.blit` on standard arrays unless they are known to not be flat float arrays. (There remains a direct call in `Dummy.Array.extend`, but the type tells us that the source array is one of our own non-flat arrays.)

Of course, this would be vastly easier if we could rely on first-class support for "uniform arrays", as proposed in https://github.com/ocaml/RFCs/pull/37 . (Or maybe, the complexity would be moved from Dynarray, where it feels accidental, to the Uniform_array module, where it makes sense to have it.) I don't have the energy to push for this right now, but hopefully we will have them at some point and be able to simplify the code here.

Reply:
Yes, the trickiness of the Dynarray code makes me even more convinced of the relevance of uniform arrays.